### PR TITLE
doc: add memory management HLD

### PR DESCRIPTION
**Pull request by dbkinder**
Transcribe and publish the reviewed memory managment HLD into the ACRN
doc set as a developer guide

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

### CODE REVIEW

**Comment by NanlinXie** [comment]
Can one of the admins verify this patch?

File: doc/developer-guides/memmgt-hld.rst [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+-  How ACRN hypervisor manages host memory (HPA/HVA)
+-  How ACRN hypervisor manages SOS guest memory (HPA/GPA)
+-  How ACRN hypervisor & SOS DM manage UOS guest memory (HPA/GPA)
+
+Hypervisor Memory Management
+****************************
+
+The ACRN hypervisor is the primary owner to manage system
+memory. Typically the BIOS/bootloader passes the platform physical
```
Comment (gvancuts) [inline doc/developer-guides/memmgt-hld.rst]: Would it be more appropriate to talk about "boot firmware (e.g. EFI)" as we don't support legacy Bios.

File: doc/developer-guides/memmgt-hld.rst [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+memory layout - E820 table. The ACRN hypervisor does its memory
+management based on this table.
+
+Physical Memory Layout - E820
+=============================
+
+The BIOS/bootloader pass the E820 table through a multiboot protocol.
```
Comment (gvancuts) [inline doc/developer-guides/memmgt-hld.rst]: Passes?

File: doc/developer-guides/memmgt-hld.rst [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+   :width: 900px
+   :name: vm-layout
+
+   Hypervisor Virtual Memory Layout
+
+:numref:`vm-layout` shows:
+
+-  Hypervisor can access all of system memory
+-  Hypervisor has an UNCACHED MMIO/PCI hole reserved for devices, such
+   as for LAPIC/IOAPIC accessing
```
Comment (gvancuts) [inline doc/developer-guides/memmgt-hld.rst]: access instead of accessing?

File: doc/developer-guides/memmgt-hld.rst [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+   :name: sos-mem-layout
+
+   SOS Physical Memory Layout
+
+Host to Guest Mapping
+=====================
+
+ACRN hypervisor creates Service OS's host (HPA) to guest (GPA) mapping
+(EPT mapping) through the function
+prepare_vm0_memmap_and_e820() when it creates SOS VM. It follows
```
Comment (gvancuts) [inline doc/developer-guides/memmgt-hld.rst]: Should we enclose prepare_vm0_memmap_and_e820() in double-backticks?

File: doc/developer-guides/memmgt-hld.rst [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+change after the Service OS begins running. Each native device driver
+can access its MMIO through this static mapping. EPT violation is only
+serving for vLAPIC/vIOAPIC's emulation in the hypervisor for Service OS
+VM.
+
+User OS Memory Management
+*************************
+
+User OS VM is created by the DM (Device
+eMulator) application running in the Service
```
Comment (gvancuts) [inline doc/developer-guides/memmgt-hld.rst]: we usually refer to DM as Device Model, we should stick to that unless there is a specific reason not to.

File: doc/developer-guides/memmgt-hld.rst [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+OS. DM is responsible for the memory allocation for an User or Guest OS
```
Comment (gvancuts) [inline doc/developer-guides/memmgt-hld.rst]: I thought that 'user' was a bit of an exception in that it should be preceded by 'a' (and not 'an')?

**Comment by dbkinder** [comment]
thanks @gvancuts.  Comments addressed with this update.

File: doc/developer-guides/memmgt-hld.rst [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+The ACRN hypervisor is the primary owner to manage system
+memory. Typically the boot firmware (e.g., EFI) passes the platform physical
+memory layout - E820 table to the hypervisor. The ACRN hypervisor does its memory
+management based on this table.
+
+Physical Memory Layout - E820
+=============================
+
+The BIOS/bootloader passes the E820 table through a multiboot protocol.
```
Comment (gvancuts) [inline doc/developer-guides/memmgt-hld.rst]: Can you also update this to `boot firmware (e.g., EFI)`? (sorry I had missed it earlier).

**Comment by dbkinder** [comment]
done.

**Pull request merged**